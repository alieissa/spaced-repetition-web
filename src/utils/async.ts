/** @format */
/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Untriggered = {
  type: 'Untriggered'
}
export function Untriggered(): Untriggered {
  return {
    type: 'Untriggered',
  }
}

const t: Optional<Answer, 'easiness' | 'quality' | 'interval'> = {
  id: '',
  content: 'fsdfds',
}
export type Loading<V> = {
  type: 'Loading'
  value: V
}
export function Loading<V>(v: V): Loading<V> {
  return {
    type: 'Loading',
    value: v,
  }
}

export type Failure<E> = {
  type: 'Failure'
  value: E
}
export function Failure<E>(e: E): Failure<E> {
  return {
    type: 'Failure',
    value: e,
  }
}

export type Success<V> = {
  type: 'Success'
  value: V
}
export function Success<V>(v: V): Success<V> {
  return {
    type: 'Success',
    value: v,
  }
}

export type Async<L = unknown, F = unknown, S = unknown> =
  | Untriggered
  | Loading<L>
  | Failure<F>
  | Success<S>

export type AsyncStatus<R, E, A> = Async<R, E, A>['type']

export type AsyncMap<R, E, A, U> = {
  [K in AsyncStatus<R, E, A>]: U extends { type: K } ? U : never
}

export type AsyncStatusMap<R, E, A> = AsyncMap<R, E, A, Async<R, E, A>>
export type AsyncPattern<T, R, E, A> = {
  [K in keyof AsyncStatusMap<R, E, A>]: (a: AsyncStatusMap<R, E, A>[K]) => T
}

export function match<R, E, A>(
  e: Async<R, E, A>,
): <T>(pattern: AsyncPattern<T, R, E, A>) => T {
  return <T>(pattern: AsyncPattern<T, R, E, A>) => pattern[e.type](e as any)
}
