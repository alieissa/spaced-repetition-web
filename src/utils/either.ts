/** @format */

/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Right<V> = Readonly<{
  type: 'Right'
  value: V
}>
export function Right<V>(v: V): Right<V> {
  return {
    type: 'Right',
    value: v,
  }
}

export type Left<E = Error> = Readonly<{
  type: 'Left'
  value: E
}>
export function Left<E>(e: E): Left<E> {
  return {
    type: 'Left',
    value: e,
  }
}

export type Either<L, R> = Left<L> | Right<R>

export type EitherType<L, R> = Either<L, R>['type']
export type EitherMap<L, R, U> = {
  [K in EitherType<L, R>]: U extends { type: K } ? U : never
}
export type EitherTypeMap<L, R> = EitherMap<L, R, Either<L, R>>
export type EitherPattern<T, L, R> = {
  [K in keyof EitherTypeMap<L, R>]: (either: EitherTypeMap<L, R>[K]) => T
}

export function match<L, R>(
  e: Either<L, R>,
): <T>(pattern: EitherPattern<T, L, R>) => T {
  return <T>(pattern: EitherPattern<T, L, R>) => pattern[e.type](e as any)
}
