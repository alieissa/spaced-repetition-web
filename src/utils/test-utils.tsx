/** @format */

import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { RouterProvider, createMemoryRouter } from 'react-router-dom'
import decksReducer from 'src/modules/decks/decks.reducer'

// As a basic setup, import your same slice reducers
import { loginReducer } from 'src/modules/login'
import { signupReducer } from 'src/modules/signup'
import { verificationReducer } from 'src/modules/verification'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
// interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
//   preloadedState?: PreloadedState<RootState>
//   store?: AppStore
// }

const reducers = combineReducers({
  login: loginReducer,
  signup: signupReducer,
  verification: verificationReducer,
  decks: decksReducer,
})

const withRedux = (ui: React.ReactElement) => {
  const store = configureStore({ reducer: reducers })
  return <Provider store={store}>{ui}</Provider>
}
const withRouter = (ui: React.ReactElement) => {
  const router = createMemoryRouter([{ element: ui, path: '/' }], {
    initialEntries: ['/'],
    initialIndex: 1,
  })

  return <RouterProvider router={router} />
}
export function renderWithProviders(ui: React.ReactElement) {
  // Return an object with the store and all of RTL's query functions
  return render(withRedux(withRouter(ui)))
}

export const flushPromises = () =>
  new Promise((resolve) => setTimeout(resolve, 0))
