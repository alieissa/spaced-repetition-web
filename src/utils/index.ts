/** @format */

import * as async from './async'
import * as either from './either'

export { async }
export { either }
