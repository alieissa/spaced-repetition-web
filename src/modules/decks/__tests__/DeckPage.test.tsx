/** @format */

import { faker } from '@faker-js/faker'
import '@testing-library/jest-dom'
import { act, screen, waitForElementToBeRemoved } from '@testing-library/react'
import { rest } from 'msw'
import { setupServer } from 'msw/node'

import { flushPromises, renderWithProviders } from 'src/utils/test-utils'
import { DeckPage } from '..'
const mockNavigate = jest.fn(() => ({}))
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}))

const deckId = faker.string.uuid()
const decksUrl = `${process.env.REACT_APP_API_ENDPOINT}/decks/${deckId}`

// Mounting component this way ensures that the route parameter :deckId
// has id of the deck, i.e. deckId defined above
const mountComponent = () =>
  renderWithProviders(<DeckPage />, {
    initialEntries: [`/decks/${deckId}`],
    path: 'decks/:deckId',
  })

const handlers = [
  rest.get(decksUrl, (__, res, ctx) => {
    return res(
      ctx.json({
        id: deckId,
        name: faker.word.noun(),
        questions: [
          {
            id: faker.string.uuid(),
            content: faker.lorem.sentence(5),
            answers: {
              id: faker.string.uuid(),
              content: faker.lorem.sentence(5),
            },
          },
        ],
      }),
      ctx.status(200),
    )
  }),
]

const server = setupServer(...handlers)
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

describe('DeckPage', () => {
  describe('view', () => {
    it('should render correctly', () => {
      const { asFragment } = renderWithProviders(<DeckPage />)
      expect(asFragment()).toMatchSnapshot()
    })

    it('should display loading', async () => {
      mountComponent()
      const testId = 'deck-loading'
      expect(await screen.findByTestId(testId)).toBeInTheDocument()
      // See https://davidwcai.medium.com/react-testing-library-and-the-not-wrapped-in-act-errors-491a5629193b
      await waitForElementToBeRemoved(() => screen.queryByTestId(testId))
    })

    it('should display success', async () => {
      mountComponent()

      await act(flushPromises)

      const testId = 'deck-success'
      expect(await screen.findByTestId(testId)).toBeInTheDocument()
    })

    it('should display failure', async () => {
      server.use(
        rest.get(decksUrl, (__, res, ctx) => {
          return res(ctx.status(500))
        }),
      )
      mountComponent()

      await act(flushPromises)

      const testId = 'deck-failure'
      expect(await screen.findByTestId(testId)).toBeInTheDocument()
    })
  })

  // NOTE: The interaction tests will be done for the component
  // that will be common to NewDeck and Deck.
})
