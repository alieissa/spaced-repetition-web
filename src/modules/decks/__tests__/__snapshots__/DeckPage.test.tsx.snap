// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeckEdit view should render correctly 1`] = `
<DocumentFragment>
  <div
    data-testid="deck-success"
  >
    <div
      class="justify-space-between"
      style="padding: 1rem 1rem 1rem 2rem; line-height: 2rem; border-bottom: 1px solid;"
      title="Edit Deck"
    >
      <div
        class="align-center"
      >
        <div>
          <button
            class="ui huge icon button"
            style="padding-left: 0px; padding-right: 0px; background: none;"
          >
            <i
              aria-hidden="true"
              class="chevron left icon"
            />
          </button>
        </div>
        <div
          class="ui header"
          style="font-size: 1.5rem; margin-top: 0px; margin-bottom: 0px;"
        >
          Edit Deck
        </div>
      </div>
    </div>
    <main
      class="px-2r"
    >
      <div>
        <div>
          <div
            class="ui huge input w-full"
            style="opacity: 1;"
          >
            <input
              data-testid="deck-name"
              placeholder="Enter name here"
              style="border-radius: unset; font-weight: 400; padding-left: 0px;"
              type="text"
              value="Test Name"
            />
          </div>
        </div>
        <div
          class="flex-column"
          style="margin-top: 1.5rem;"
          title="Description"
        >
          <div
            class="ui small header"
            style="margin-bottom: 0px; margin-top: 0px;"
          >
            Description
          </div>
          <div>
            <textarea
              class="w-full"
              data-testid="deck-description"
              name="deck-textarea"
              rows="5"
            />
          </div>
        </div>
        <div
          class="flex-column"
          style="margin-top: 1.5rem;"
          title="Cards"
        >
          <div
            class="ui small header"
            style="margin-bottom: 0px; margin-top: 0px;"
          >
            Cards
          </div>
          <div>
            <form
              class="ui form w-full bordered"
            >
              <div
                class="flex-row-reverse"
              >
                <i
                  aria-hidden="true"
                  class="arrow up icon"
                  data-testid="view-answers-btn"
                />
              </div>
              <div
                class="ui horizontal list flex"
                role="list"
                style="display: flex;"
              >
                <div
                  class="item flex-1"
                  role="listitem"
                  style="padding: 0px;"
                >
                  <div
                    class="error field w-full"
                  >
                    <div
                      class="ui input"
                      style="opacity: 1;"
                    >
                      <input
                        aria-invalid="true"
                        data-testid="question-content"
                        placeholder="Enter question here"
                        style="border-radius: unset;"
                        type="text"
                        value="Test Question"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="item flex-1"
                  role="listitem"
                  style="padding: 0px;"
                >
                  <div
                    class="ui list"
                    role="list"
                    style="padding: 0px;"
                  >
                    <div
                      class="item flex"
                      role="listitem"
                      style="display: flex; padding: 0px;"
                    >
                      <div
                        class="error field w-full"
                      >
                        <div
                          class="ui input"
                          style="opacity: 1;"
                        >
                          <input
                            aria-invalid="true"
                            data-testid="answer-content-0"
                            placeholder="Enter answer here"
                            style="border-radius: unset;"
                            type="text"
                            value="Test Content"
                          />
                        </div>
                      </div>
                      <button
                        class="ui small icon disabled button"
                        disabled=""
                        style="padding-left: 0px; padding-right: 0px; background: none;"
                        tabindex="-1"
                      >
                        <i
                          aria-hidden="true"
                          class="x icon"
                        />
                      </button>
                    </div>
                    <div
                      class="item"
                      role="listitem"
                      style="text-align: right; padding: 0px;"
                    >
                      <button
                        class="ui small icon button"
                        style="padding-left: 0px; padding-right: 0px; background: none;"
                      >
                        <i
                          aria-hidden="true"
                          class="plus icon"
                        />
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </form>
            <div
              class="ui basic segment flex-row-reverse"
              style="padding: 0px;"
            >
              <button
                class="ui green icon button"
                style="padding-left: 0px; padding-right: 0px; background: none;"
              >
                <i
                  aria-hidden="true"
                  class="green plus icon"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <footer
        class="flex-row-reverse pb-2r"
      >
        <button
          class="ui green small button"
          data-testid="deck-save"
          style="border-radius: 0;"
          type="submit"
        >
          Save
        </button>
      </footer>
    </main>
  </div>
</DocumentFragment>
`;
